pkg/kubectlcmd/client.go

```diff
@@ -47,92 +47,171 @@ func (console) Run(c *exec.Cmd) error {
+	extraArgs ...string) (string, string, error) {
+	subcmds := []string{"get", "cm", name}
+	params := &kubectlParams{
+		dryRun:     false,
+		verbose:    false,
+		kubeconfig: kubeconfig,
+		context:    context,
+		namespace:  namespace,
+		stdin:      "",
+		output:     output,
+		extraArgs:  extraArgs,
+	}
+	return c.kubectl(subcmds, params)
+}
 
-	return stdout.String(), csError, nil
+func logAndPrint(v ...interface{}) {
+	s := fmt.Sprintf(v[0].(string), v[1:]...)
+	log.Infof(s)
+	fmt.Println(s)
 }
 
-// GetConfig runs the kubectl get cm command with the provided argument
-func (c *Client) GetConfig(name, namespace, output string, extraArgs ...string) (string, string, error) {
-	args := []string{"get", "cm", name}
-	if namespace != "" {
-		args = append(args, "-n", namespace)
+// Apply runs the `kubectl` command by specifying subcommands in subcmds with parameters:
+// dryRun to not actually run the cmd, verbose to not print stdin
+// kubeconfig, context to identify a cluster,
+// namespace to locate a k8s namespace, and
+// returns stdout, stderr, error getting from running this `kubectl` command.
+func (c *Client) kubectl(subcmds []string, params *kubectlParams) (string, string, error) {
+	hasStdin := strings.TrimSpace(params.stdin) != ""
+	args := subcmds
+	if params.kubeconfig != "" {
+		args = append(args, "--kubeconfig", params.kubeconfig)
+	}
+	if params.context != "" {
+		args = append(args, "--context", params.context)
 	}
-	if output != "" {
-		args = append(args, "-o", output)
+	if params.namespace != "" {
+		args = append(args, "-n", params.namespace)
 	}
-	args = append(args, extraArgs...)
+	if params.output != "" {
+		args = append(args, "-o", params.output)
+	}
+	args = append(args, params.extraArgs...)
 
-	cmd := exec.Command("kubectl", args...)
+	if hasStdin {
+		args = append(args, "-f", "-")
+	}
 
-	var stdout bytes.Buffer
-	var stderr bytes.Buffer
+	cmd := exec.Command("kubectl", args...)
+	var stdout, stderr bytes.Buffer
 	cmd.Stdout = &stdout
 	cmd.Stderr = &stderr
 
+	cmdStr := strings.Join(args, " ")
+	if hasStdin {
+		cmd.Stdin = strings.NewReader(params.stdin)
+		if params.verbose {
+			cmdStr += "\n" + params.stdin
+		} else {
+			cmdStr += " <use --verbose to see stdin string> \n"
+		}
+	}
+
+	if params.dryRun {
+		logAndPrint("dry run mode: would be running this cmd:\n%s\n", cmdStr)
+		return "", "", nil
+	}
+
+	log.Infof("running command:\n%s\n", cmdStr)
 	err := c.cmdSite.Run(cmd)
+	csError := util.ConsolidateLog(stderr.String())
+
 	if err != nil {
-		logAndPrint("error running kubectl get cm: %s", err)
-		return stdout.String(), stderr.String(), fmt.Errorf("error running kubectl %s: %s", strings.Join(args, " "), err)
+		logAndPrint("error running kubectl: %s", err)
+		return stdout.String(), csError, fmt.Errorf("error running kubectl: %s", err)
 	}
 
-	logAndPrint("kubectl get cm success")
-
-	return stdout.String(), stderr.String(), nil
-}
+	logAndPrint("kubectl success")
```

Comment:
probably better to have this just as a log.Info

Reply:
Good point! It is actually annoying when it comes to users.

```
...
Parent dependency for Prometheus has unblocked, proceeding.
kubectl applying manifest for component Citadel
kubectl applying manifest for component Prometheus
kubectl applying manifest for component Galley
kubectl applying manifest for component Pilot
kubectl applying manifest for component IngressGateway
kubectl applying manifest for component Policy
kubectl applying manifest for component Injector
kubectl success
kubectl success
kubectl success
kubectl success
kubectl success
kubectl success
kubectl success
kubectl success
kubectl success
kubectl success
kubectl success
kubectl success
```

Reply:
Fixed.